source/agora/consensus/data/TransactionData.d

```diff
@@ -0,0 +1,297 @@
+                    return cast(ubyte)(10 + c - 'a');
+                case 'A': .. case 'F':
+                    return cast(ubyte)(10 + c - 'A');
+                default:
+                    return 0xff;    //  mark error
+            }
+        };
+
+        ubyte [] data;
+        data.length = str.length / 2;
+
+        size_t idx;
+        foreach (chunk; str.byChar.map!(c => hexDecoder(c)).chunks(2))
+        {
+            if (chunk[0] == 0xff || chunk[1] == 0xff)
+                return TransactionData([], "The hexadecimal string contains an abnormal character");
+
+            data[idx++] = cast(ubyte)((chunk[0] << 4) + chunk[1]);
+        }
+
+        return TransactionData(data);
+    }
+
+    /***************************************************************************
+
+        Implements hashing support
+
+        Params:
+            dg = Hashing function accumulator
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @nogc @safe
+    {
+        hashPart(this.data, dg);
+    }
+
+    /***************************************************************************
+
+        TransactionData Serialization
+
+        Params:
+            dg = serialize function accumulator
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @safe
+    {
+        serializePart(this.data, dg);
+    }
+
+    /***************************************************************************
+
+        Returns a new instance of type `TransactionData`
+
+        Params:
+            TransactionDataT = Qualified type of TransactionData to return
+            dg   = Delegate to read binary data
+            opts = Deserialization options (should be forwarded)
+
+        Returns:
+            A new instance of type `TransactionData`
+
+    ***************************************************************************/
+
+    public static TransactionDataT fromBinary (TransactionDataT = TransactionData) (
+        scope DeserializeDg dg, const ref DeserializerOptions opts) @safe
+    {
+        return TransactionData(deserializeFull!(ubyte[])(dg, opts));
+    }
+
+    /// Support for comparison
+    public int opCmp (ref const typeof(this) s) const nothrow @safe @nogc
+    {
+        size_t len = min(this.length, s.length);
+        foreach (idx; 0 .. len)
+            if (this.data[idx] != s.data[idx])
+                return this.data[idx] - s.data[idx];
+
+        return (this.length == s.length) ? 0
+                                         : ((this.length > s.length) ? 1 : -1);
+    }
+
+    /// Support for comparison (rvalue overload)
+    public int opCmp (const typeof(this) s) const nothrow @safe @nogc
+    {
+        return this.opCmp(s);
+    }
+
+    /// Returns the number of elements in the data
+    @property size_t length() const pure nothrow @safe @nogc
```

Comment:
`this.data` is already public, so why do you need this getter function?

Reply:
There is no particular reason, so I will remove it